test/daml/Test/Lifecycle.daml

```diff
@@ -25,67 +31,230 @@ type C = Claim Observation Date Decimal Text
+
+  setDate today
+  t <- getDate
+
+  Lifecycle.Result{remaining, pending} <- Lifecycle.settle observe25 (bond [today, tomorrow, afterTomorrow]) t
+  remaining === bond [tomorrow, afterTomorrow]
+  pending === [(15.0, a)]
+
+  setDate tomorrow
+  t <- getDate
+
+  Lifecycle.Result{remaining, pending} <- Lifecycle.settle observe25 remaining t
+  remaining === bond [afterTomorrow]
+  pending === [(15.0, a)]
+
+  -- Check bond coupon doesn't get processed twice
+  Lifecycle.Result{remaining, pending} <- Lifecycle.settle observe25 remaining t
+  remaining === bond [afterTomorrow]
+  pending === []
+
+  setDate afterTomorrow
+  t <- getDate
+
+  Lifecycle.Result{remaining, pending} <- Lifecycle.settle observe25 remaining t
   remaining === Zero
-  pending === pure (1.0, a)
+  pending === [(15.0, a), (1_000.0, a)]
+
+testEuropeanCall = script do
+  let strike = 23.0
+      payoff = Scale (O.observe "spot" - O.pure strike) (One a)
+      option = european tomorrow payoff
+
+  setDate today
+  t <- getDate
+
+  remaining <- Lifecycle.exercise observe25 Zero option t
+  remaining === option
 
-  -- For github issue #10
-  let outerScale : C = Scale (O.pure 2.0) $ When (at $ date 2021 Mar 11) (One a)
-  Lifecycle.Result{remaining, pending} <-
-    Lifecycle.lifecycle spot chooseLeft outerScale (date 1970 Jan 1)
-  remaining === outerScale
-  pending === mempty
-  Lifecycle.Result{remaining, pending} <-
-    Lifecycle.lifecycle spot chooseLeft outerScale (date 2021 Mar 11)
+  Lifecycle.Result{remaining, pending} <- Lifecycle.settle observe25 option t
+  remaining === option
+  pending === []
+
+  setDate tomorrow
+  t <- getDate
+
+  remaining <- Lifecycle.exercise observe25 payoff option t
+  remaining === payoff
+
+  Lifecycle.Result{remaining, pending} <- Lifecycle.settle observe25 remaining t
   remaining === Zero
   pending === pure (2.0, a)
+
+-- Knock-in tomorrow
+testAmericanPut = script do
+  let strike = 30.0
+      payoff = Scale (O.pure strike - O.observe "spot") (One a)
+      option = american tomorrow afterTomorrow payoff
+
+  -- Scenario 1) Before knock-in
+  setDate today
+  t <- getDate
+
+  Lifecycle.Result{remaining, pending} <- Lifecycle.settle observe25 option t
+  remaining === option
+  pending === []
+
+  remaining <- Lifecycle.exercise observe25 payoff remaining t
+  remaining === option -- not in acquisition range yet
+
+  -- Scenario 2) in acquisition range
+  setDate tomorrow
+  t <- getDate
+
+  -- Settlement before exercise has been done is a no-op
+  Lifecycle.Result{remaining, pending} <- Lifecycle.settle observe25 option t
+  remaining === Until (O.TimeGte afterTomorrow) (Anytime (O.TimeGte tomorrow) (payoff `Or` Zero))
+  pending === []
+  
+  remaining <- Lifecycle.exercise observe25 (payoff `Or` Zero) remaining t
+  remaining === Until (O.TimeGte afterTomorrow) (payoff `Or` Zero)
+
+  remaining <- Lifecycle.exercise observe25 payoff remaining t
+  remaining === Until (O.TimeGte afterTomorrow) payoff
```

Comment:
It's a bit annoying here that we need to lifecycle twice, once to remove the `Anytime` wrapper and a second time to remove the not-selected `Or` branch.
On the other hand, if say I would instead take as an argument the 'desired end state' (i.e. `payoff` in this case), the client would need to implement complicated logic to derive this; in the current form, they just take one of the child nodes of `Anytime` or `Or`; that's trivial.